pass environment variables from config file to plugins and steps

balintzs opened the pull request:
* config file path defined in env variable: `SLG_PLUGIN_ENV_CONFIG`
* load config properties into a dictionary grouped by sections that correspond to docker images

skygeo commented on slingshot/core/step.py:
```diff
@@ -24,16 +26,38 @@ class StepStatus(Enum):
 
+    # Dictionary of the lists of common and image-specific
+    # environment variables
+    CONFIGS = _load_env_configs(os.getenv("SLG_PLUGIN_ENV_CONFIG"))
```
I would advice not loading the list of environment (aka _load_env_configs) in this file and only rely on having the data passed to the class through the __init__ method. I would then load the list of environments in cli.py, and pass the data down the chain from cli to server to engine to runner to step.
This for tow main reasons:
- consistency: currently all the application specific settings are configured in the cli.py file. It would be sad to start moving away from this pattern
- it kind of breaks the class encapsulation: using public method to pass parameters seems a far better choice.

skygeo replied:
On a similar topic than my previous comment, instead of loading part of the config (the list of environments) from a file, part of the config from environment variables, could we look into centralizing all the slingshot engine config into one config file? (and remove the clicks parameters)

For example, starting the server would be 

```
slingshot server -c config.ini
```
```
# config.ini
[default]
backend=swf # this value defaults to local
verbosity=DEBUG # this value defaults to INFO 

[environment]
foo=bar
bar=baz
[swf]
domain=slingshot
version=1.01
```

balintzs replied:
If you pass these configs via the `__init__` method then they need to be stored in the engine. Furthermore, if we want these values to be passed onto the actual workflow steps, the engine needs to individually override each step created by the plugin, increasing complexity a lot. Keeping it in the step, we make sure that these configs are only used by the appropriate class. Having said that, I'm not against loading these configs in the cli in a command line argument.

balintzs replied:
Those configurations are for different tools: some for the engine, some for the plugins and steps. I won't mix them up.

chaudyg replied:
Can we get more opinion on that? This will make it difficult to split the workflow engine from the deployment engine!

balintzs replied:
If you look at the implementation now, the configs are loaded in CLI and set using a class method of the `Step`. This makes it easy to separate the two things.

skygeo replied:
Had a quick chat with @gjtempleton to gather more opionion. 
A single place for all the configuration makes a lot of sense. I think we want to strip-out the click options and provide a single config file with everything. This is a fairly common pattern (docker.cfg, nginx.cfg, haproxy.cfg, ...)
Note that we are having the same challenge in the flyby project, so we should find a solution that is consistent accross the board.

skygeo replied:
Following a quick chat with @gjtempleton .
I am not comfortable with the having a class variable with state shared across every step instances. Class variables are sometime useful, but in this case it make the logic hard to debug the code hard to re-use (we are stating that every container we start have this list of environment variable...it's a bit harsh). I would rather pass the list of environments as an argument of the Step  constructor method. Then we can be explicit, and decide if this list of environment are passed to the plugins, or the steps, or everything.

balintzs commented:
Added changes to load environment configs in CLI and pass them to the `Step` class. The format also changed, so that we use yaml like everywhere else.

balintzs commented:
OK, so apparently every change here is wrong. You can implement it the way you want. I'm closing this.
